Class Tweakers: Introduction & Access Widening (#517)
FabricMC/fabric-docs: develop/classtweaker/index.md

---
title: Class Tweakers
description: Learn what class tweakers are, and how to set them up.
authors:
  - cassiancc
  - Earthcomputer
  - its-miroma
  - MildestToucan
---

> Class tweakers, known as access wideners prior to gaining further functionality, are used to complement Mixin's bytecode manipulation by providing additional transformation tools and allowing for certain runtime modifications to be visible from the development environment.

Reviewer: ```suggestion
Class tweakers, formerly known as access wideners before gaining further functionality, provide transformation tools complementary to Mixin bytecode manipulation. They also allow some runtime modifications to be accessible within the development environment.
```

Class tweakers are not specific to a given Minecraft version, but are only available starting from Fabric Loader 0.18.0, and may only target Vanilla Minecraft classes.

## Setup {#setup}

### File Format {#file-format}

> Class tweaker files are conventionally named after your modid, appended with `.classtweaker` to help IDE plugins recognize them. They should be stored in `resources`.

Reviewer: ```suggestion
Class tweaker files are conventionally named after your mod ID, `example-mod.classtweaker`, to help IDE plugins recognize them. They should be stored in `resources`.
```

The file must have the following header as its first line:

```txt
classTweaker v1 named
```

Class tweaker files can have blank lines and comments starting with `#`. Comments can start at the end of a line.

Whilst the specific syntax depends on the feature, modifications are each declared on separate lines. An entry's elements can be separated using any whitespace, including tabs.

### Specifying The File Location {#specifying-the-file-location}

The class tweaker file's location must be specified in your `build.gradle` and `fabric.mod.json` files. Remember that you must also depend on Fabric Loader 0.18.0 or above to use class tweakers.

> The specifications are still named after access wideners for backward compatibility reasons.

Reviewer: ```suggestion
```

the docs don't usually document history

Reviewer 2: This case feels important to me as it's not obvious where the name comes from otherwise.

Reviewer: Alright

```suggestion

The specifications are still named after access wideners to preserve backwards compatibility.
```

I never heard "backward compatibility" at the singular, yet Wikipedia says it's correct...
